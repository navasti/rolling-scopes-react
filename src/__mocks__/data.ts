import { ErrorMessages } from 'appConstants';
import {
  PokemonMoveDetails,
  PokemonTypeDetails,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  PokemonData,
  PokemonMove,
  PokemonType,
  Details,
  Pokemon,
  Lengths,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const links = ['Home', 'About', 'Form', 'Invalid'];

export const testingComponentName = 'Test';
export const testingLocation = '/test';

export const testingContent = 'Content';
export const testingTitle = 'Title';

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

//

export const lengths: Lengths = { pokemons: 10, moves: 15, types: 20 };

const samplePokemon = (): PokemonDetails => ({
  base_experience: 62,
  name: 'charmander',
  weight: 85,
  height: 6,
  id: 4,
  abilities: [
    {
      ability: { name: 'blaze', url: 'https://pokeapi.co/api/v2/ability/66/' },
      is_hidden: false,
      slot: 1,
    },
  ],
  sprites: {
    front_default: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png',
    back_default:
      'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/4.png',
  },
  stats: [
    {
      base_stat: 52,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
  ],
  types: [
    {
      slot: 1,
      type: {
        name: 'fire',
        url: 'https://pokeapi.co/api/v2/type/10/',
      },
    },
  ],
});

const sampleMove = (): PokemonMoveDetails => ({
  accuracy: 100,
  name: 'tackle',
  priority: 0,
  power: 40,
  id: 33,
  pp: 35,
  learned_by_pokemon: [
    {
      name: 'bulbasaur',
      url: 'https://pokeapi.co/api/v2/pokemon/1/',
    },
  ],
  type: {
    name: 'normal',
    url: 'https://pokeapi.co/api/v2/type/1/',
  },
});

const sampleType = (): PokemonTypeDetails => ({
  moves: [{ name: 'gust', url: 'https://pokeapi.co/api/v2/move/16/' }],
  name: 'flying',
  id: 3,
  damage_relations: {
    double_damage_from: [
      { name: 'electric', url: 'https://pokeapi.co/api/v2/type/13/' },
      { name: 'rock', url: 'https://pokeapi.co/api/v2/type/6/' },
    ],
    double_damage_to: [{ name: 'fighting', url: 'https://pokeapi.co/api/v2/type/2/' }],
    half_damage_from: [{ name: 'bug', url: 'https://pokeapi.co/api/v2/type/7/' }],
    half_damage_to: [{ name: 'steel', url: 'https://pokeapi.co/api/v2/type/9/' }],
    no_damage_from: [{ name: 'ground', url: 'https://pokeapi.co/api/v2/type/5/' }],
    no_damage_to: [],
  },
  pokemon: [
    { pokemon: { name: 'charizard', url: 'https://pokeapi.co/api/v2/pokemon/6/' }, slot: 2 },
  ],
});

const prepareBuilder = <T extends Details>(sample: () => T) => {
  const builder = (object = sample()) => ({
    withName: (name: T['name']) => builder({ ...object, name }),
    withId: (id: T['id']) => builder({ ...object, id }),
    valueOf: () => object,
  });
  return builder;
};

const pokemonBuilder = prepareBuilder(samplePokemon);
const moveBuilder = prepareBuilder(sampleMove);
const typeBuilder = prepareBuilder(sampleType);

export const pokemonsMock: Array<PokemonDetails> = [
  pokemonBuilder().withId(1).withName('bulbasaur'),
  pokemonBuilder().withId(25).withName('pikachu'),
  pokemonBuilder(),
].map((builder) => builder.valueOf());

export const movesMock: Array<PokemonMoveDetails> = [
  moveBuilder().withId(55).withName('water-gun'),
  moveBuilder().withId(50).withName('disable'),
  moveBuilder(),
].map((builder) => builder.valueOf());

export const typesMock: Array<PokemonTypeDetails> = [
  typeBuilder().withId(2).withName('fighting'),
  typeBuilder().withId(5).withName('ground'),
  typeBuilder(),
].map((builder) => builder.valueOf());
