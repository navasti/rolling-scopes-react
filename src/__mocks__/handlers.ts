import { screen } from '@testing-library/react';
import { AvailableTabs } from 'appConstants';

export const expectByText = jest.fn((text: string) => {
  expect(screen.getByText(text)).toBeInTheDocument();
});

export const handleCloseModal = jest.fn();
export const handleCardSelect = jest.fn();
export const handleOpenModal = jest.fn();
export const onTabClick = jest.fn((availableTab: AvailableTabs) => availableTab);
