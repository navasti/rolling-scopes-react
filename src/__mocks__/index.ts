export { testOnChange, handleCloseModal } from './handlers';
export { localStorageMock } from './localStorage';

export {
  fetchBaseMock,
  fetchDetailsMock,
  handleResponseMock,
  fetchByParameterMock,
  handleMappedResponseMock,
} from './api';

export {
  links,
  pokemon,
  visible,
  invisible,
  isLoading,
  movesMock,
  typesMock,
  testErrors,
  pokemonData,
  pokemonMove,
  pokemonType,
  testingTitle,
  pokemonsMock,
  customPokemon,
  testingContent,
  testingLocation,
  testEmptyErrors,
  testingComponentName,
  moveModalContentProperties,
  typeModalContentProperties,
  pokemonModalContentProperties,
} from './data';

export {
  modalRef,
  errorRef,
  inputRef,
  selectRef,
  ModalTitle,
  ModalContent,
  TestingElement,
  TestingErrorElement,
} from './elements';
