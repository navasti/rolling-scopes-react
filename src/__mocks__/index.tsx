import { ErrorMessages } from 'appConstants';
import { createRef, RefObject } from 'react';
import { hasError } from 'utils';
import {
  MockDetailedDataType,
  PokemonMoveDetails,
  PokemonTypeDetails,
  MockBaseDataType,
  PokemonDetails,
  CustomPokemon,
  ErrorsObject,
  CheckFields,
  PokemonData,
  PokemonMove,
  PokemonType,
  TextFields,
  Pokemon,
} from 'types';

export const visible = true;
export const invisible = false;
export const isLoading = false;

export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const links = ['Home', 'About', 'Form', 'Invalid'];
export const componentName = 'About';
export const location = '/about';
export const testingContent = 'Content';
export const testingTitle = 'Title';

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const TestingTitleElement = () => <span>{testingTitle}</span>;
export const TestingContentElement = () => <div>{testingContent}</div>;
export const modalRef: RefObject<HTMLDivElement> = createRef();

export const fetchBaseMock = <T extends MockBaseDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve({ results: [mockData] }),
    })
  ) as jest.Mock;
};

export const fetchByParameterMock = <T extends MockDetailedDataType>(mockData: T) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData),
    })
  ) as jest.Mock;
};

export const fetchDetailsMock = <T,>(mockData: Array<T>) => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockData[0]),
    })
  ) as jest.Mock;
};

export const pokemonType: PokemonType = {
  name: 'typex',
  url: 'url',
};

export const pokemon: Pokemon = {
  name: 'pokemonix',
  url: 'url',
};

export const pokemonMove: PokemonMove = {
  name: 'movix',
  url: 'url',
};

export const typeModalContentProperties = [
  'Double damage from',
  'Double damage to',
  'Half damage from',
  'Half damage to',
  'No damage from',
  'No damage to',
  'Pokemons',
  'Moves',
] as const;

export const moveModalContentProperties = [
  'Learned by',
  'Priority',
  'Accuracy',
  'Power',
  'Type',
  'PP',
] as const;

export const pokemonModalContentProperties = [
  'Pokemon statistics',
  'Pokemon abilities',
  'Base experience',
  'Height',
  'Weight',
  'Types',
] as const;

export const pokemonData: PokemonData = {
  results: [{ name: 'testName', url: 'testURL' }],
  previous: null,
  next: null,
  count: 1,
};

export const testOnChange = (field: TextFields | CheckFields) => console.log(field);

export const testEmptyErrors: ErrorsObject = {
  birthday: null,
  consent: null,
  gender: null,
  name: null,
  type: null,
};

export const testErrors: ErrorsObject = {
  birthday: ErrorMessages.birthday,
  consent: ErrorMessages.consent,
  gender: ErrorMessages.gender,
  name: ErrorMessages.name,
  type: ErrorMessages.type,
};

export const customPokemon: CustomPokemon = {
  birthday: '01-01-1999',
  name: 'pokemonix',
  gender: 'male',
  avatar: null,
  type: 'fire',
  shiny: true,
};

export const detailedTypes: Array<PokemonTypeDetails> = [
  {
    damage_relations: {
      double_damage_from: [],
      double_damage_to: [{ name: 'toTest', url: 'urlTest' }],
      half_damage_from: [{ name: 'fromTest', url: 'urlTest' }],
      half_damage_to: [],
      no_damage_from: [],
      no_damage_to: [],
    },
    pokemon: [{ pokemon: { name: 'pokemonTest', url: 'urlTest' }, slot: 1 }],
    moves: [{ name: 'moveTest', url: 'urlTest' }],
    name: 'typeTest',
    id: 10,
  },
];

export const detailedMoves: Array<PokemonMoveDetails> = [
  {
    learned_by_pokemon: [{ name: 'pokomonTest', url: 'urlTest' }],
    type: { name: 'typeTest', url: 'urlTest' },
    name: 'moveTest',
    accuracy: 60,
    priority: 0,
    power: 50,
    id: 20,
    pp: 15,
  },
];

export const detailedPokemons: Array<PokemonDetails> = [
  {
    abilities: [{ ability: { name: 'ability' }, is_hidden: false, slot: 2 }],
    stats: [{ base_stat: 33, effort: 33, stat: { name: 'test' } }],
    sprites: { back_default: 'url', front_default: 'url' },
    types: [{ slot: 2, type: { name: 'typeTest' } }],
    name: 'Testing pokemon',
    base_experience: 12,
    height: 15,
    weight: 10,
    id: 34,
  },
  {
    abilities: [{ ability: { name: 'ability2' }, is_hidden: true, slot: 1 }],
    stats: [{ base_stat: 22, effort: 22, stat: { name: 'test' } }],
    sprites: { back_default: 'url2', front_default: 'url2' },
    types: [{ slot: 2, type: { name: 'typeTest2' } }],
    name: 'Testing pokemon2',
    base_experience: 16,
    height: 11,
    weight: 11,
    id: 35,
  },
];

type Store = Record<string, string>;

export const localStorageMock = (() => {
  let store: Store = {};
  return {
    getItem(key: string) {
      return store[key];
    },
    setItem(key: string, value: string) {
      store[key] = value.toString();
    },
    clear() {
      store = {};
    },
    removeItem(key: string) {
      delete store[key];
    },
  };
})();
