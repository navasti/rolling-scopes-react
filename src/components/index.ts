export { Loader } from './Loader';
export { Modal } from './Modal';
