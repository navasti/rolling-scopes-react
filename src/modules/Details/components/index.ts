export { MoveCardDetails } from './MoveCardDetails';
export { TypeCardDetails } from './TypeCardDetails';
export { PokemonCardDetails } from './PokemonCardDetails';
