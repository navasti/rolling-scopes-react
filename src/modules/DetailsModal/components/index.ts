export { TypeModalContent } from './TypeModalContent';
export { MoveModalContent } from './MoveModalContent';
export { PokemonModalContent } from './PokemonModalContent';
