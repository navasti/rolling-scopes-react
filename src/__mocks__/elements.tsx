import { createRef, RefObject } from 'react';
import { testingContent, testingTitle } from './data';

export const modalRef: RefObject<HTMLDivElement> = createRef();
export const errorRef: RefObject<HTMLSpanElement> = createRef();
export const inputRef: RefObject<HTMLInputElement> = createRef();
export const selectRef: RefObject<HTMLSelectElement> = createRef();

export const TestingElement = () => <p>testing element</p>;
export const TestingErrorElement = () => <p>testing error</p>;

export const ModalTitle = () => <span data-testid="modal-title">{testingTitle}</span>;
export const ModalContent = () => <div data-testid="modal-content">{testingContent}</div>;
