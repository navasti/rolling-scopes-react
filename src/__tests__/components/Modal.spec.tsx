import { fireEvent, render, screen } from '@testing-library/react';
import { handleCloseModal } from '__mocks__/handlers';
import { ModalContent } from '__mocks__/elements';
import { testingTitle } from '__mocks__/data';
import { Modal } from 'components';

const prepareModalElement = (opened: boolean) => (
  <Modal handleCloseModal={handleCloseModal} opened={opened} title={testingTitle}>
    <ModalContent />
  </Modal>
);

describe('Modal', () => {
  beforeAll(() => {
    const div = document.createElement('div');
    div.id = 'modal-root';
    div.dataset.testid = 'modal-mock';
    document.body.appendChild(div);
  });
  afterAll(() => handleCloseModal.mockReset());
  it('modal should be rendered when open prop equals true', () => {
    render(
      <>
        <Modal handleCloseModal={handleCloseModal} opened={true} title={testingTitle}>
          <ModalContent />
        </Modal>
      </>
    );
    expect(screen.getByText(testingTitle)).toBeInTheDocument();
    expect(screen.getByTestId('modal-content')).toBeInTheDocument();
  });
  it('modal should not be rendered when open prop equals false', () => {
    const { container } = render(prepareModalElement(false));
    expect(screen.queryByTestId('modal-content')).toBeNull();
    expect(screen.queryByTestId('modal-title')).toBeNull();
    expect(container).toBeEmptyDOMElement();
  });
  it('close button should be rendered and clicking it should fire handleCloseModal method', () => {
    render(prepareModalElement(true));
    const modal = screen.getByTestId('modal-mock');
    const button = modal.querySelector('button') as HTMLButtonElement;
    expect(button).toBeInTheDocument();
    expect(handleCloseModal).not.toHaveBeenCalled();
    fireEvent.click(button);
    expect(handleCloseModal).toHaveBeenCalledTimes(1);
  });
  it('clicking outside the modal should fire handleCloseModal method', () => {
    render(prepareModalElement(true));
    const modal = screen.getByTestId('modal-mock');
    expect(handleCloseModal).not.toHaveBeenCalled();
    fireEvent.click(modal.children[0]);
    expect(handleCloseModal).toHaveBeenCalledTimes(1);
  });
  it('modal should match the snapshot', () => {
    render(prepareModalElement(true));
    expect(screen.getByTestId('modal-mock')).toMatchSnapshot();
  });
});
