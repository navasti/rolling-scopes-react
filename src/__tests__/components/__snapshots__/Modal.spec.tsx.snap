// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal modal should match the snapshot 1`] = `
<div
  data-testid="modal-mock"
  id="modal-root"
>
  <div
    class="sc-gsnTZi kOjvcg"
  >
    <div
      class="sc-dkzDqf jCvLml"
    >
      <header
        class="sc-hKMtZM cuhbOU"
      >
        <p
          class="sc-gKXOVf iuBSNt"
        >
          Title
        </p>
        <button
          class="sc-eCYdqJ iPyUzY"
        >
          <svg>
            times.svg
          </svg>
        </button>
      </header>
      <div
        class="sc-jSMfEi ihWpvY"
      >
        <div
          data-testid="modal-content"
        >
          Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
